pareto/utilities/process_data.py

```diff
@@ -0,0 +1,932 @@
+            default_used.append(param_tab)
+    # Raise warning for default values
+    if len(default_used) > 0:
+        warning_message = (
+            f"The following parameters were missing and default values were substituted:"
+            + str(default_used)
+        )
+        warnings.warn(warning_message, stacklevel=3)
+    return (df_sets, df_parameters)
+
+
+def model_infeasibility_detection(strategic_model):
+    # check_required_data() performs basic feasibility checks:
+    # - at least one source node in the model
+    # - at least on sink node in the model
+    # - each source node has an outgoing arc, each sink node has an incoming arc, and all other nodes have both an incoming and an outgoing arc.
+
+    capacity_feasibility_message = []
+    demand_feasibility_message = []
+    # Checks that for each time period, the volume of produced water is below the maximum capacity limit of sink nodes
+    # Get the total system produced water for each time period (PadRates, FLowbackRates)
+    strategic_model.p_beta_TotalPW = Param(
+        strategic_model.s_T,
+        units=strategic_model.model_units["volume"],
+        doc="Combined water supply forecast (flowback & production) over the planning horizon [volume]",
+        mutable=True,  # Mutable Param - can be changed in sensitivity analysis without rebuilding the entire model
+    )
+
+    # Get the maximum capacity of sink nodes for each time period
+    # Sink nodes: CompletionsDemand, (InitialDisposalCapacity + max(disposalCapacityIncrements)) * DisposalOperatingCapacity,
+    # StorageEvaporation, ReuseCapacity, InitialStorageCapacity + Storage CapacityIncrements,
+    # (InitialTreatmentCapacity+TreatmentCapacityIncrements)*TreatmentResidual
+    strategic_model.p_sigma_MaxPWCapacity = Param(
+        strategic_model.s_T,
+        units=strategic_model.model_units["volume"],
+        doc="Combined produced water capacity in system [volume]",
+        mutable=True,  # Mutable Param - can be changed in sensitivity analysis without rebuilding the entire model
+    )
+
+    strategic_model.p_gamma_TimePeriodDemand = Param(
+        strategic_model.s_T,
+        units=strategic_model.model_units["volume"],
+        doc="Total water demand over the planning horizon [volume]",
+        mutable=True,  # Mutable Param - can be changed in sensitivity analysis without rebuilding the entire model
+    )
+
+    strategic_model.p_sigma_WaterAvailable = Param(
+        strategic_model.s_T,
+        units=strategic_model.model_units["volume"],
+        doc="Total PW and external water available [volume]",
+        mutable=True,  # Mutable Param - can be changed in sensitivity analysis without rebuilding the entire model
+    )
+
+    for t in strategic_model.s_T:
+        strategic_model.p_beta_TotalPW[t] = sum(
+            strategic_model.p_beta_Production[p, t]
+            + strategic_model.p_beta_Flowback[p, t]
+            for p in strategic_model.s_P
+        )
```

Comment:
E.g., this could be implemented as a Pyomo expression.

Reply:
Done!